orchagent/natorch.cpp

```diff
@@ -0,0 +1,4691 @@
+         m_countersDb(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0),
+         m_countersNatTable(&m_countersDb, COUNTERS_NAT_TABLE),
+         m_countersNaptTable(&m_countersDb, COUNTERS_NAPT_TABLE),
+         m_countersTwiceNatTable(&m_countersDb, COUNTERS_TWICE_NAT_TABLE),
+         m_countersTwiceNaptTable(&m_countersDb, COUNTERS_TWICE_NAPT_TABLE),
+         m_countersGlobalNatTable(&m_countersDb, COUNTERS_GLOBAL_NAT_TABLE),
+         m_stateWarmRestartEnableTable(stateDb, STATE_WARM_RESTART_ENABLE_TABLE_NAME),
+         m_stateWarmRestartTable(stateDb, STATE_WARM_RESTART_TABLE_NAME),
+         m_natQueryTable(appDb, APP_NAT_TABLE_NAME),
+         m_naptQueryTable(appDb, APP_NAPT_TABLE_NAME),
+         m_twiceNatQueryTable(appDb, APP_NAT_TWICE_TABLE_NAME),
+         m_twiceNaptQueryTable(appDb, APP_NAPT_TWICE_TABLE_NAME),
+         nullIpv4Addr(0)
+{
+    /* Set NAT admin mode to disabled */
+    admin_mode = "disabled";
+
+    /* Set NAT default timeout as 600 seconds */
+    timeout = 600;
+
+    /* Set NAT default tcp timeout as 86400 seconds (1 Day) */
+    tcp_timeout = 86400;
+
+    /* Set NAT default udp timeout as 300 seconds */
+    udp_timeout = 300;
+
+    /* Set entries count to 0 */
+    totalEntries = totalSnatEntries = totalDnatEntries = 0;
+    totalStaticNatEntries = totalDynamicNatEntries = 0;
+    totalStaticNaptEntries = totalDynamicNaptEntries = 0;
+    totalStaticTwiceNatEntries = totalDynamicTwiceNatEntries = 0;
+    totalStaticTwiceNaptEntries = totalDynamicTwiceNaptEntries = 0;
+ 
+    /* Add NAT notifications support from APPL_DB */
+    SWSS_LOG_INFO("Add NAT notifications support from APPL_DB ");
+    m_flushNotificationsConsumer = new NotificationConsumer(appDb, "FLUSHNATREQUEST");
+    auto flushNotifier = new Notifier(m_flushNotificationsConsumer, this, "FLUSHNATREQUEST");
+    Orch::addExecutor(flushNotifier);
+
+    SWSS_LOG_INFO("Add REDIS DB cleanup notification support");
+    m_cleanupNotificationConsumer = new NotificationConsumer(appDb, "NAT_DB_CLEANUP_NOTIFICATION");
+    auto cleanupNotifier = new Notifier(m_cleanupNotificationConsumer, this, "NAT_DB_CLEANUP_NOTIFICATION");
+    Orch::addExecutor(cleanupNotifier);
+
+    /* Start the timer to query NAT entry statistics every 5 secs and hitbits every 30 secs */
+    SWSS_LOG_INFO("Start the HITBIT Timer ");
+    auto interval      = timespec { .tv_sec = NAT_HITBIT_N_CNTRS_QUERY_PERIOD, .tv_nsec = 0 };
+    m_natQueryTimer = new SelectableTimer(interval);
+    auto executor   = new ExecutableTimer(m_natQueryTimer, this, "NAT_HITBIT_N_CNTRS_QUERY_TIMER");
+    Orch::addExecutor(executor);
+
+    /* Get the Maximum supported entries */
+    SWSS_LOG_INFO("Get the Maximum supported SNAT entries");
+    sai_status_t     status;
+    sai_attribute_t  attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.id = SAI_SWITCH_ATTR_AVAILABLE_SNAT_ENTRY;
+    maxAllowedNatEntries = 0;
+
+    status = sai_switch_api->get_switch_attribute(gSwitchId, 1, &attr);
```

Comment:
please move this SAI call to when NAT is enabled from CLI

Reply:
Get operation on a unimplemented SAI attribute doesn't result in crash, so retaining in current changes as discussed.